architecture-decisions/0003-preservation.md

```diff
@@ -0,0 +1,84 @@
+# 3. Preservation
+
+Date: 2019-04-02
+
+## Status
+
+Accepted
+
+## Context
+
+We have agreed that we will preserve digital objects by saving resources in
+Google Cloud Storage in a directory structure which preserves both the binaries
+the resource is made up of as well as the JSON serialization of the resource
+itself.
+
+## Decisions
+
+1. Preserving
+   1. We will preserve materials in Google Cloud Coldline Storage with
+      `versioning` enabled. Versions will be kept indefinitely and without
+      limit. All files will go in a single bucket.
+   1. When a resource is `complete` and marked with the `cloud` preservation
+      policy it will save itself and all resources contained in `member_ids` in
+      a directory structure in Google Cloud Storage that looks like the
+      following:
+      ```
+      - <resource-id>
+        - data
+          - <child-id>
+            - <child-id>.json
+            - <binary.tif>
+        - <resource-id>.json
+      ```
+   1. When a child with a preserved parent (any length up the hierarchy) is
+      `complete` and marked with the `cloud` preservation policy and is updated,
+      it will update itself.
+   1. Related objects such as collections, Ephemera Terms, etc. will not be
+      packaged inside the preserved object. If it's important they be preserved,
+      those objects should be marked with the `cloud` preservation policy.
+   1. When a FileSet is added to a resource which is already complete and marked
+      with the `cloud` preservation policy, it will upload the new binary
+      content to both the repository and to Google Cloud STorage
+   1. If a child is marked to be preserved, but its parent is not, it will still
+      save in a nested directory structure, but will not automatically create
+      backups of its parents.
+   1. This behavior will be attached to the ChangeSetPersister.
+1. Packaging Details
+   1. Resources created in cloud storage will be attached to the appropriate
+      resource in `file_metadata` with a PCDM Use value of
+      `pcdm:PreservationMasterFile`
+   1. We will store the checksums of files uploaded to cloud storage in the
+      `file_metadata` attached to the owning resource.
+   1. We will only keep the most recent version of any file, overwriting any
+      files which match the same file name, but relying on versioning to go back
+      if necessary.
+   1. When a preserved resource is deleted, we will delete its directory from
+      preservation storage. If we need to get it again, we will look at Google
+      Cloud Storage's stored versions.
+
+## Consequences
+
+1. Structure
+   1. The resultant structure will not be in a format that is expected by
+      outside vendors. However, we have a BagIt packager for those use cases.
```

Comment:
Do we want to say something here about being a subset of the BagIt spec?  I.e., you could take this format and add the manifests to make it a valid bag.